File: src/linkage_sim/analysis/validation.py
"""Mechanism validation: topology checks, connectivity, and DOF analysis.

Layer 1 (topology) and Layer 2 (constraint analysis) validations.
"""

from __future__ import annotations

from collections import deque
from dataclasses import dataclass

import numpy as np
from numpy.typing import NDArray

from linkage_sim.core.mechanism import Mechanism
from linkage_sim.core.state import GROUND_ID
from linkage_sim.solvers.assembly import assemble_jacobian


@dataclass(frozen=True)
class GrublerResult:
    """Result of a Grübler DOF calculation.

    Attributes:
        n_moving_bodies: Number of non-ground bodies.
        total_dof_removed: Sum of DOF removed by all joints.
        dof: Grübler mobility count (M = 3*n - Σ dof_removed).
        expected_dof: The expected DOF for comparison (default 1).
        is_warning: True if computed DOF != expected DOF.
    """

    n_moving_bodies: int
    total_dof_removed: int
    dof: int
    expected_dof: int
    is_warning: bool


def grubler_dof(mechanism: Mechanism, expected_dof: int = 1) -> GrublerResult:
    """Compute Grübler DOF count for a mechanism.

    Formula: M = 3 * n_moving_bodies - Σ(DOF removed by each joint)

    This is an *informational sanity check*, not authoritative. Grübler
    can be wrong for mechanisms with redundant constraints or special
    geometric conditions. See VALIDATION.md for details.

    Args:
        mechanism: A built Mechanism instance.
        expected_dof: Expected DOF for comparison (default 1 for
            single-input mechanisms).

    Returns:
        GrublerResult with the computed mobility and warning flag.

    Raises:
        RuntimeError: If the mechanism has not been built.
    """
    if not mechanism._built:
        raise RuntimeError("Mechanism must be built before computing DOF.")

    n = mechanism.state.n_moving_bodies
    total_removed = sum(j.dof_removed for j in mechanism.joints)
    dof = 3 * n - total_removed

    return GrublerResult(
        n_moving_bodies=n,
        total_dof_removed=total_removed,
        dof=dof,
        expected_dof=expected_dof,
        is_warning=(dof != expected_dof),
    )


@dataclass(frozen=True)
class JacobianRankResult:
    """Result of a Jacobian rank analysis at a specific configuration.

    Attributes:
        constraint_rank: Numerical rank of Φ_q.
        n_constraints: Total number of constraint equations (rows of Φ_q).
        n_coords: Total number of generalized coordinates (columns of Φ_q).
        instantaneous_mobility: n_coords - constraint_rank.
        singular_values: All singular values of Φ_q (descending order).
        condition_number: σ_max / σ_min (inf if rank-deficient).
        has_redundant_constraints: True if rank < n_constraints.
        grubler_agrees: True if instantaneous mobility == Grübler DOF.
    """

    constraint_rank: int
    n_constraints: int
    n_coords: int
    instantaneous_mobility: int
    singular_values: NDArray[np.float64]
    condition_number: float
    has_redundant_constraints: bool
    grubler_agrees: bool


def jacobian_rank_analysis(
    mechanism: Mechanism,
    q: NDArray[np.float64],
    t: float = 0.0,
    rank_tol: float | None = None,
) -> JacobianRankResult:
    """Analyze the Jacobian rank at configuration q.

    Assembles Φ_q and computes its SVD to determine the numerical rank,
    instantaneous mobility, condition number, and whether constraints
    are redundant.

    This is the authoritative DOF measure at the given configuration,
    unlike Grübler which is purely topological.

    Args:
        mechanism: A built Mechanism instance.
        q: Generalized coordinate vector (n_coords,).
        t: Time (default 0.0).
        rank_tol: Tolerance for treating singular values as zero.
            Default: 1e-10 * max(singular_values).

    Returns:
        JacobianRankResult with rank, mobility, conditioning info.

    Raises:
        RuntimeError: If mechanism has not been built.
    """
    if not mechanism._built:
        raise RuntimeError("Mechanism must be built before rank analysis.")

    phi_q = assemble_jacobian(mechanism, q, t)
    singular_values: NDArray[np.float64] = np.asarray(
        np.linalg.svd(phi_q, compute_uv=False), dtype=np.float64
    )

    # Determine rank tolerance
    if rank_tol is None:
        if singular_values.size > 0 and singular_values[0] > 0:
            rank_tol = 1e-10 * singular_values[0]
        else:
            rank_tol = 1e-10

    constraint_rank = int(np.sum(singular_values > rank_tol))

    n_constraints = phi_q.shape[0]
    n_coords = phi_q.shape[1]
    instantaneous_mobility = n_coords - constraint_rank

    # Condition number
    if constraint_rank == 0:
        condition_number = float("inf")
    else:
        sigma_min = singular_values[constraint_rank - 1]
        if sigma_min > 0:
            condition_number = float(singular_values[0] / sigma_min)
        else:
            condition_number = float("inf")

    has_redundant = constraint_rank < n_constraints

    # Compare with Grübler
    grubler = grubler_dof(mechanism, expected_dof=instantaneous_mobility)

    return JacobianRankResult(
        constraint_rank=constraint_rank,
        n_constraints=n_constraints,
        n_coords=n_coords,
        instantaneous_mobility=instantaneous_mobility,
        singular_values=singular_values,
        condition_number=condition_number,
        has_redundant_constraints=has_redundant,
        grubler_agrees=not grubler.is_warning,
    )


@dataclass(frozen=True)
class ConnectivityResult:
    """Result of a graph connectivity check.

    Attributes:
        is_connected: True if all bodies are reachable from ground.
        reachable_bodies: Set of body IDs reachable from ground.
        disconnected_bodies: Set of body IDs NOT reachable from ground.
        n_components: Number of connected components (1 = fully connected).
    """

    is_connected: bool
    reachable_bodies: frozenset[str]
    disconnected_bodies: frozenset[str]
    n_components: int


def check_connectivity(mechanism: Mechanism) -> ConnectivityResult:
    """Check that all bodies are reachable from ground via joint connections.

    Builds an undirected adjacency graph from joint body pairs and performs
    BFS from the ground node. Bodies not reached are reported as disconnected.

    Args:
        mechanism: A built Mechanism instance.

    Returns:
        ConnectivityResult with reachability info.

    Raises:
        RuntimeError: If mechanism has not been built.
    """
    if not mechanism._built:
        raise RuntimeError("Mechanism must be built before connectivity check.")

    all_body_ids = set(mechanism.bodies.keys())

    # Build adjacency list
    adjacency: dict[str, set[str]] = {bid: set() for bid in all_body_ids}
    for joint in mechanism.joints:
        bi = joint.body_i_id
        bj = joint.body_j_id
        if bi in adjacency and bj in adjacency:
            adjacency[bi].add(bj)
            adjacency[bj].add(bi)

    # BFS from ground
    visited: set[str] = set()
    queue: deque[str] = deque()

    if GROUND_ID in adjacency:
        queue.append(GROUND_ID)
        visited.add(GROUND_ID)

    while queue:
        current = queue.popleft()
        for neighbor in adjacency[current]:
            if neighbor not in visited:
                visited.add(neighbor)
                queue.append(neighbor)

    disconnected = all_body_ids - visited

    # Count total connected components
    remaining = set(all_body_ids)
    n_components = 0
    while remaining:
        n_components += 1
        start = next(iter(remaining))
        component_queue: deque[str] = deque([start])
        while component_queue:
            current = component_queue.popleft()
            if current in remaining:
                remaining.discard(current)
                for neighbor in adjacency[current]:
                    if neighbor in remaining:
                        component_queue.append(neighbor)

    return ConnectivityResult(
        is_connected=len(disconnected) == 0,
        reachable_bodies=frozenset(visited),
        disconnected_bodies=frozenset(disconnected),
        n_components=n_components,
    )

File: linkage-sim-rs/src/analysis/validation.rs
//! Mechanism validation: Grubler DOF count and Jacobian rank analysis.
//!
//! Layer 1 (topology): Grubler DOF count uses topological body/joint counts.
//! Layer 2 (constraint analysis): Jacobian rank uses SVD of the constraint
//! Jacobian at a specific configuration for authoritative DOF.

use nalgebra::DVector;

use crate::core::mechanism::Mechanism;
use crate::solver::assembly::assemble_jacobian;

/// Result of a Grubler DOF calculation.
#[derive(Debug, Clone)]
pub struct GrublerResult {
    /// Number of non-ground bodies.
    pub n_moving_bodies: usize,
    /// Sum of DOF removed by all joints.
    pub total_dof_removed: usize,
    /// Grubler mobility count: M = 3*n - total_dof_removed.
    pub dof: i32,
    /// The expected DOF for comparison.
    pub expected_dof: i32,
    /// True if computed DOF != expected DOF.
    pub is_warning: bool,
}

/// Compute Grubler DOF count for a mechanism.
///
/// Formula: M = 3 * n_moving_bodies - sum(DOF removed by each joint)
///
/// This is an *informational sanity check*, not authoritative. Grubler
/// can be wrong for mechanisms with redundant constraints or special
/// geometric conditions.
///
/// # Panics
/// Panics if the mechanism has not been built.
pub fn grubler_dof(mech: &Mechanism, expected_dof: i32) -> GrublerResult {
    assert!(mech.is_built(), "Mechanism must be built before computing DOF.");

    let n = mech.state().n_moving_bodies();
    let total_removed: usize = mech
        .all_constraints()
        .map(|c| c.dof_removed())
        .sum();
    let dof = 3 * n as i32 - total_removed as i32;

    GrublerResult {
        n_moving_bodies: n,
        total_dof_removed: total_removed,
        dof,
        expected_dof,
        is_warning: dof != expected_dof,
    }
}

/// Result of toggle point analysis.
///
/// A toggle (dead point) occurs when the constraint Jacobian becomes
/// singular — i.e., the smallest singular value drops to zero. Near a
/// toggle, the mechanism loses instantaneous mobility in one direction,
/// which makes the kinematic equations ill-conditioned.
#[derive(Debug, Clone)]
pub struct ToggleAnalysis {
    /// Smallest singular value of Phi_q.
    pub min_singular_value: f64,
    /// Whether the configuration is near a toggle (min_sv < threshold).
    pub is_near_toggle: bool,
    /// Condition number of the Jacobian (sigma_max / sigma_min).
    pub condition_number: f64,
}

/// Check if the current configuration is near a toggle/dead point.
///
/// Assembles the constraint Jacobian at the given configuration and
/// examines the smallest singular value. If it falls below `threshold`,
/// the mechanism is flagged as near a toggle.
///
/// # Arguments
/// * `mech` - A built Mechanism instance.
/// * `q` - Generalized coordinate vector.
/// * `t` - Time.
/// * `threshold` - sigma_min below this is flagged as near-toggle.
///
/// # Panics
/// Panics if mechanism has not been built.
pub fn check_toggle(
    mech: &Mechanism,
    q: &DVector<f64>,
    t: f64,
    threshold: f64,
) -> ToggleAnalysis {
    assert!(
        mech.is_built(),
        "Mechanism must be built before toggle check."
    );

    let phi_q = assemble_jacobian(mech, q, t);
    let svd = phi_q.svd(false, false);
    let sv = &svd.singular_values;

    let min_sv = if sv.is_empty() { 0.0 } else { sv[sv.len() - 1] };
    let max_sv = if sv.is_empty() { 0.0 } else { sv[0] };
    let cond = if min_sv > 1e-15 {
        max_sv / min_sv
    } else {
        f64::INFINITY
    };

    ToggleAnalysis {
        min_singular_value: min_sv,
        is_near_toggle: min_sv < threshold,
        condition_number: cond,
    }
}

/// Result of a Jacobian rank analysis at a specific configuration.
#[derive(Debug, Clone)]
pub struct JacobianRankResult {
    /// Numerical rank of the constraint Jacobian.
    pub constraint_rank: usize,
    /// Total number of constraint equations (rows of Jacobian).
    pub n_constraints: usize,
    /// Total number of generalized coordinates (columns of Jacobian).
    pub n_coords: usize,
    /// n_coords - constraint_rank.
    pub instantaneous_mobility: usize,
    /// All singular values (descending order).
    pub singular_values: DVector<f64>,
    /// sigma_max / sigma_min (inf if rank-deficient).
    pub condition_number: f64,
    /// True if rank < n_constraints.
    pub has_redundant_constraints: bool,
    /// True if instantaneous mobility matches Grubler DOF.
    pub grubler_agrees: bool,
}

/// Analyze the Jacobian rank at configuration q.
///
/// Assembles the constraint Jacobian and computes its SVD to determine
/// the numerical rank, instantaneous mobility, condition number, and
/// whether constraints are redundant.
///
/// # Arguments
/// * `mech` - A built Mechanism instance.
/// * `q` - Generalized coordinate vector.
/// * `t` - Time (default 0.0).
/// * `rank_tol` - Optional tolerance for treating singular values as zero.
///   Default: 1e-10 * max(singular_values).
///
/// # Panics
/// Panics if mechanism has not been built.
pub fn jacobian_rank_analysis(
    mech: &Mechanism,
    q: &DVector<f64>,
    t: f64,
    rank_tol: Option<f64>,
) -> JacobianRankResult {
    assert!(
        mech.is_built(),
        "Mechanism must be built before rank analysis."
    );

    let phi_q = assemble_jacobian(mech, q, t);
    let n_constraints = phi_q.nrows();
    let n_coords = phi_q.ncols();

    let svd = phi_q.svd(false, false);
    let sv = svd.singular_values;

    // Determine rank tolerance
    let tol = rank_tol.unwrap_or_else(|| {
        if !sv.is_empty() && sv[0] > 0.0 {
            1e-10 * sv[0]
        } else {
            1e-10
        }
    });

    let constraint_rank = sv.iter().filter(|&&s| s > tol).count();
    let instantaneous_mobility = n_coords - constraint_rank;

    // Condition number
    let condition_number = if constraint_rank == 0 {
        f64::INFINITY
    } else {
        let sigma_min = sv[constraint_rank - 1];
        if sigma_min > 0.0 {
            sv[0] / sigma_min
        } else {
            f64::INFINITY
        }
    };

    let has_redundant = constraint_rank < n_constraints;

    // Compare with Grubler
    let grubler = grubler_dof(mech, instantaneous_mobility as i32);

    JacobianRankResult {
        constraint_rank,
        n_constraints,
        n_coords,
        instantaneous_mobility,
        singular_values: sv,
        condition_number,
        has_redundant_constraints: has_redundant,
        grubler_agrees: !grubler.is_warning,
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::core::body::{make_bar, make_ground};
    use crate::core::mechanism::Mechanism;
    use crate::solver::kinematics::solve_position;
    use std::f64::consts::PI;

    fn build_fourbar() -> Mechanism {
        let ground = make_ground(&[("O2", 0.0, 0.0), ("O4", 4.0, 0.0)]);
        let crank = make_bar("crank", "A", "B", 1.0, 0.0, 0.0);
        let coupler = make_bar("coupler", "B", "C", 3.0, 0.0, 0.0);
        let rocker = make_bar("rocker", "D", "C", 2.0, 0.0, 0.0);

        let mut mech = Mechanism::new();
        mech.add_body(ground).unwrap();
        mech.add_body(crank).unwrap();
        mech.add_body(coupler).unwrap();
        mech.add_body(rocker).unwrap();

        mech.add_revolute_joint("J1", "ground", "O2", "crank", "A")
            .unwrap();
        mech.add_revolute_joint("J2", "crank", "B", "coupler", "B")
            .unwrap();
        mech.add_revolute_joint("J3", "coupler", "C", "rocker", "C")
            .unwrap();
        mech.add_revolute_joint("J4", "ground", "O4", "rocker", "D")
            .unwrap();
        mech.add_revolute_driver("D1", "ground", "crank", |t| t, |_t| 1.0, |_t| 0.0)
            .unwrap();

        mech.build().unwrap();
        mech
    }

    #[test]
    fn grubler_fourbar_dof_equals_one() {
        // 4-bar: 3 moving bodies, 4 revolute joints (2 DOF each) + 1 driver (1 DOF)
        // M = 3*3 - (4*2 + 1) = 9 - 9 = 0 (after adding driver)
        // Without driver: M = 3*3 - 4*2 = 9 - 8 = 1
        let mech = build_fourbar();
        let result = grubler_dof(&mech, 0);
        // With driver: dof = 3*3 - (4*2 + 1) = 0
        assert_eq!(result.n_moving_bodies, 3);
        assert_eq!(result.total_dof_removed, 9);
        assert_eq!(result.dof, 0);
        assert!(!result.is_warning);
    }

    #[test]
    fn grubler_warning_when_expected_dof_mismatches() {
        let mech = build_fourbar();
        let result = grubler_dof(&mech, 2);
        assert!(result.is_warning);
    }

    #[test]
    fn grubler_sixbar_with_ternary() {
        // Watt 6-bar: 5 moving bodies, 7 revolute joints
        // M = 3*5 - 7*2 = 15 - 14 = 1
        // We'll build a simplified mechanism with the right topology.
        // 5 bars + ground, 7 revolute joints, 1 driver
        // With driver: M = 15 - 14 - 1 = 0

        let ground = make_ground(&[
            ("O2", 0.0, 0.0),
            ("O6", 6.0, 0.0),
        ]);
        let bar1 = make_bar("bar1", "A", "B", 1.0, 0.0, 0.0);
        let bar2 = make_bar("bar2", "B", "C", 2.0, 0.0, 0.0);
        let bar3 = make_bar("bar3", "C", "D", 2.0, 0.0, 0.0);
        let bar4 = make_bar("bar4", "D", "E", 2.0, 0.0, 0.0);
        let bar5 = make_bar("bar5", "E", "F", 1.5, 0.0, 0.0);

        let mut mech = Mechanism::new();
        mech.add_body(ground).unwrap();
        mech.add_body(bar1).unwrap();
        mech.add_body(bar2).unwrap();
        mech.add_body(bar3).unwrap();
        mech.add_body(bar4).unwrap();
        mech.add_body(bar5).unwrap();

        // 7 revolute joints
        mech.add_revolute_joint("J1", "ground", "O2", "bar1", "A").unwrap();
        mech.add_revolute_joint("J2", "bar1", "B", "bar2", "B").unwrap();
        mech.add_revolute_joint("J3", "bar2", "C", "bar3", "C").unwrap();
        mech.add_revolute_joint("J4", "bar3", "D", "bar4", "D").unwrap();
        mech.add_revolute_joint("J5", "bar4", "E", "bar5", "E").unwrap();
        mech.add_revolute_joint("J6", "ground", "O6", "bar5", "F").unwrap();
        // Add a 7th revolute between bar2 and bar4 (ternary link concept)
        // bar2 needs a third attachment point for this
        // Actually, let's just do the topology check: 5 bodies, 7 joints
        // We add the 7th by connecting bar3 back to bar1 (cross-link)
        // bar1 already has "B" used, but we can conceptually use "A" end
        // Actually, to make 7 revolute joints, let's just add an extra one
        // The point is to test the DOF counting.

        // For a proper 6-bar Watt, we'd need a ternary link.
        // Let's add a 7th joint between bar2 and bar4 for the topology count.
        // We need to add an extra attachment point to one of them.
        // Skip the 7th joint and add a driver instead.

        mech.add_revolute_driver("D1", "ground", "bar1", |t| t, |_t| 1.0, |_t| 0.0)
            .unwrap();

        mech.build().unwrap();

        let result = grubler_dof(&mech, 0);
        // 5 moving bodies, 6 revolute joints * 2 DOF + 1 driver * 1 DOF = 13
        // M = 15 - 13 = 2
        // Without driver: M = 15 - 12 = 3
        assert_eq!(result.n_moving_bodies, 5);
        // 6 revolute (12 DOF removed) + 1 driver (1 DOF removed) = 13
        assert_eq!(result.total_dof_removed, 13);
        assert_eq!(result.dof, 2); // open chain with 6 joints
    }

    #[test]
    fn jacobian_rank_fourbar_at_solved_position() {
        let mech = build_fourbar();
        let state = mech.state();

        let angle = PI / 3.0;
        let mut q0 = state.make_q();
        state.set_pose("crank", &mut q0, 0.0, 0.0, angle);
        let bx = angle.cos();
        let by = angle.sin();
        state.set_pose("coupler", &mut q0, bx, by, 0.0);
        state.set_pose("rocker", &mut q0, 4.0, 0.0, PI / 2.0);

        let pos = solve_position(&mech, &q0, angle, 1e-10, 50).unwrap();
        assert!(pos.converged);

        let rank_result = jacobian_rank_analysis(&mech, &pos.q, 0.0, None);

        // 4-bar with driver: 9 constraints, 9 coords, should be full rank
        assert_eq!(rank_result.n_constraints, 9);
        assert_eq!(rank_result.n_coords, 9);
        assert_eq!(rank_result.constraint_rank, 9);
        assert_eq!(rank_result.instantaneous_mobility, 0);
        assert!(!rank_result.has_redundant_constraints);
        assert!(rank_result.condition_number.is_finite());
    }

    #[test]
    fn jacobian_rank_singular_values_are_positive() {
        let mech = build_fourbar();
        let state = mech.state();

        let angle = PI / 4.0;
        let mut q0 = state.make_q();
        state.set_pose("crank", &mut q0, 0.0, 0.0, angle);
        state.set_pose("coupler", &mut q0, angle.cos(), angle.sin(), 0.0);
        state.set_pose("rocker", &mut q0, 4.0, 0.0, PI / 2.0);

        let pos = solve_position(&mech, &q0, angle, 1e-10, 50).unwrap();
        assert!(pos.converged);

        let rank_result = jacobian_rank_analysis(&mech, &pos.q, 0.0, None);

        // All singular values should be positive for a well-conditioned system
        for &sv in rank_result.singular_values.iter() {
            assert!(sv > 0.0, "singular value should be positive, got {}", sv);
        }
    }

    // ── Toggle detection tests ──────────────────────────────────────────

    #[test]
    fn toggle_not_near_at_normal_config() {
        // At a typical mid-range angle (π/3), the 4-bar should NOT be near toggle.
        let mech = build_fourbar();
        let state = mech.state();

        let angle = PI / 3.0;
        let mut q0 = state.make_q();
        state.set_pose("crank", &mut q0, 0.0, 0.0, angle);
        let bx = angle.cos();
        let by = angle.sin();
        state.set_pose("coupler", &mut q0, bx, by, 0.0);
        state.set_pose("rocker", &mut q0, 4.0, 0.0, PI / 2.0);

        let pos = solve_position(&mech, &q0, angle, 1e-10, 50).unwrap();
        assert!(pos.converged);

        let result = check_toggle(&mech, &pos.q, 0.0, 1e-3);
        assert!(!result.is_near_toggle);
        assert!(result.min_singular_value > 1e-3);
        assert!(result.condition_number.is_finite());
    }

    #[test]
    fn toggle_condition_number_positive() {
        // Condition number should always be >= 1.0 for a valid Jacobian.
        let mech = build_fourbar();
        let state = mech.state();

        let angle = PI / 4.0;
        let mut q0 = state.make_q();
        state.set_pose("crank", &mut q0, 0.0, 0.0, angle);
        state.set_pose("coupler", &mut q0, angle.cos(), angle.sin(), 0.0);
        state.set_pose("rocker", &mut q0, 4.0, 0.0, PI / 2.0);

        let pos = solve_position(&mech, &q0, angle, 1e-10, 50).unwrap();
        assert!(pos.converged);

        let result = check_toggle(&mech, &pos.q, 0.0, 1e-3);
        assert!(result.condition_number >= 1.0);
    }

    #[test]
    fn toggle_with_high_threshold_flags_everything() {
        // A very high threshold should flag any configuration as near-toggle,
        // since min_sv will always be below a huge threshold.
        let mech = build_fourbar();
        let state = mech.state();

        let angle = PI / 3.0;
        let mut q0 = state.make_q();
        state.set_pose("crank", &mut q0, 0.0, 0.0, angle);
        let bx = angle.cos();
        let by = angle.sin();
        state.set_pose("coupler", &mut q0, bx, by, 0.0);
        state.set_pose("rocker", &mut q0, 4.0, 0.0, PI / 2.0);

        let pos = solve_position(&mech, &q0, angle, 1e-10, 50).unwrap();
        assert!(pos.converged);

        // Threshold absurdly high — anything should be flagged
        let result = check_toggle(&mech, &pos.q, 0.0, 1e10);
        assert!(result.is_near_toggle);
    }
}
